src/metrics-middleware.js

```diff
@@ -10,46 +10,73 @@ const setupOptions = {};
 
 module.exports = (appVersion, projectName, framework = 'express') => {
     return (options = {}) => {
-        const { metricsPath, defaultMetricsInterval = 10000, durationBuckets, requestSizeBuckets, responseSizeBuckets, useUniqueHistogramName, metricsPrefix, excludeRoutes, includeQueryParams } = options;
+        const {
+            metricsPath,
+            defaultMetricsInterval = 10000,
+            durationBuckets,
+            requestSizeBuckets,
+            responseSizeBuckets,
+            useUniqueHistogramName,
+            metricsPrefix,
+            excludeRoutes,
+            includeQueryParams,
+            metricsExtraLabels = [],
+            getMetricsExtraLabelValues
+        } = options;
         debug(`Init metrics middleware with options: ${JSON.stringify(options)}`);
         setupOptions.metricsRoute = metricsPath || '/metrics';
         setupOptions.excludeRoutes = excludeRoutes || [];
         setupOptions.includeQueryParams = includeQueryParams;
         setupOptions.defaultMetricsInterval = defaultMetricsInterval;
+        setupOptions.metricsExtraLabels = metricsExtraLabels || [];
+        setupOptions.getMetricsExtraLabelValues = typeof getMetricsExtraLabelValues === 'function' ? getMetricsExtraLabelValues : () => ({});
 
-        let metricNames = {
-            http_request_duration_seconds: 'http_request_duration_seconds',
-            app_version: 'app_version',
-            http_request_size_bytes: 'http_request_size_bytes',
-            http_response_size_bytes: 'http_response_size_bytes',
-            defaultMetricsPrefix: ''
-        };
-        metricNames = utils.getMetricNames(metricNames, useUniqueHistogramName, metricsPrefix, projectName);
+        const metricNames = utils.getMetricNames(
+            {
+                http_request_duration_seconds: 'http_request_duration_seconds',
+                app_version: 'app_version',
+                http_request_size_bytes: 'http_request_size_bytes',
+                http_response_size_bytes: 'http_response_size_bytes',
+                defaultMetricsPrefix: ''
+            },
+            useUniqueHistogramName,
+            metricsPrefix,
+            projectName
+        );
 
         Prometheus.collectDefaultMetrics({ timeout: defaultMetricsInterval, prefix: `${metricNames.defaultMetricsPrefix}` });
 
         PrometheusRegisterAppVersion(appVersion, metricNames.app_version);
 
+        const metricLabels = [
+            'method',
+            'route',
+            'code',
+            ...Array.isArray(metricsExtraLabels) ? metricsExtraLabels : []
```

Comment:
same as my comment on `getMetricsExtraLabelValues`, this could hide a potential bug.

i would just take `metricsExtraLabels` as it is and if it's not an array the dev would get an error.

the only problem is that spreading a string also generates an array, so maybe in this case it would be better to add input validation.

personally, i think that all input should be validated, but we can start with both `metricsExtraLabels` and `getMetricsExtraLabelValues` for now

Reply:
sounds good, just added validation to the inputs